xds/server.go

```diff
@@ -317,18 +322,66 @@ func (s *GRPCServer) GracefulStop() {
 	}
 }
 
+func routeAndProcess(ctx context.Context) error {
+	conn := transport.GetConnection(ctx)
+	cw, ok := conn.(*server.ConnWrapper)
+	if !ok {
+		return errors.New("missing connection wrapper in incoming context")
+	}
+	mn, ok := grpc.Method(ctx)
+	if !ok {
+		return errors.New("missing method name in incoming context")
+	}
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return errors.New("missing metadata in incoming context")
+	}
+	// Assuming to just take the first value defined.
+	authority := md.Get(":authority")
+	vh := xdsclient.FindBestMatchingVirtualHostServer(authority[0], cw.VirtualHosts)
+	if vh == nil {
+		return status.Error(codes.Unavailable, "the incoming RPC did not match a configured Virtual Host")
+	}
+
+	var rwi *xdsclient.RouteWithInterceptors
+	for _, r := range vh.Routes {
+		if r.M.Match(iresolver.RPCInfo{
+			Context: ctx,
+			Method:  mn,
+		}) {
+			// "NonForwardingAction is expected for all Routes used on server-side; a route with an inappropriate action causes
+			// RPCs matching that route to fail with UNAVAILABLE." - A36
+			if r.RouteAction != xdsclient.RouteActionNonForwardingAction {
+				return status.Error(codes.Unavailable, "the incoming RPC matched to a route that was not of action type non forwarding")
+			}
+			rwi = &r
```

Comment:
Should there be a `break` after this?

Reply:
The reason the tests did not catch this is I never induced a situation where this would create logical incorrectness, which would be an RPC that matches to a route action of correct type, then also afterward matches to one of incorrect type, which would then fail the RPC even though it shouldn't have. This code path was being induced (checked coverage), it just didn't have the edge case/testing support for that edge case to trigger it.

Reply:
Do you think adding a test case makes sense, now that we have spent so much time on this :)